### PR TITLE
feat: allow users to redeem positions from invalid markets

### PR DESCRIPTION
**Pull request by berteotti**
image

Closes: #104

### CODE REVIEW

File: app/components/CardBet.tsx [hunk 1]
```diff
@@ -104,11 +114,13 @@ export const CardBet = ({ userBets, children }: CardBetProps) => {
...
-                  <p>{formatValueWithFixedDecimals(outcomeBalance)}</p>
-                ) : (
-                  <p>{formatEtherWithFixedDecimals(collateralAmountSpent, 2)}</p>
-                )}
+                <p>
+                  {!market.isClosed || isWinner
+                    ? formatValueWithFixedDecimals(outcomeBalance)
+                    : market.isAnswerInvalid
+                      ? formatValueWithFixedDecimals(invalidMarketColleteralToRedeem, 2)
+                      : formatEtherWithFixedDecimals(collateralAmountSpent, 2)}
```
Comment (Diogomartf) [inline app/components/CardBet.tsx:122]: I think it's happening too much here, I'm okay with letting this here for now but we should find a better place for this logic than the the **jsx**

File: entities/markets/trade.ts [hunk 2]
```diff
@@ -36,6 +36,15 @@ export const tradesCollateralAmountSpent = ({ fpmmTrades }: TradesProps) => {
...
 
+export const tradesOutcomeTokensTraded = ({ fpmmTrades }: TradesProps) => {
+  return (
+    fpmmTrades?.reduce((acc, trade) => {
+      const type = trade.type;
+      return valueByTradeBigInt[type](acc, BigInt(trade.outcomeTokensTraded));
+    }, BigInt(0)) ?? BigInt(0)
+  );
+};
+
```
Comment (Diogomartf) [inline entities/markets/trade.ts:47]: I don't think `tradesOutcomeTokensTraded` is a good name.

It return the total, right? Not the array of trades. So I think we should add `total` or `sum` to the name.

Maybe something like `outcomeTokensTradedTotal` would describe better

File: app/components/UserBets.tsx [hunk 3]
```diff
@@ -202,16 +221,21 @@ export const UserBets = ({ fixedProductMarketMaker }: UserBets) => {
                         )}
                       >
                         <span>
-                          {tradedBalance} {collateralToken.symbol}
+                          {marketModel.isAnswerInvalid
+                            ? invalidMarketColleteralToRedeem
+                            : !canClaim && isResolved
+                              ? collateralSpent
+                              : tradedBalance}{' '}
+                          {collateralToken.symbol}
```
Comment (ElRodrigote) [inline app/components/UserBets.tsx:229]: [MINOR] [NON-BLOCKING]

I have the same opinionated and personal take here, feel free to use or dismiss the suggestion:

```tsx
function getResultAmount() {
  if (marketModel.isAnswerInvalid) return invalidMarketColleteralToRedeem;
  if (!canClaim && isResolved) return collateralSpent;

  return tradedBalance;
}

const resultAmount = getResultAmount();
```

And then inside the TSX side of the code we could do:
```tsx
<span>
  {resultAmount} {collateralToken.symbol}
</span>
```

File: app/components/UserBets.tsx [hunk 4]
```diff
@@ -239,6 +263,34 @@ export const UserBets = ({ fixedProductMarketMaker }: UserBets) => {
                     </div>
                   </>
                 )}
+                {marketModel.isAnswerInvalid && canRedeem && (
+                  <>
```
Comment (ElRodrigote) [inline app/components/UserBets.tsx:267]: [MINOR] [NON-BLOCKING]

I think it's safe to remove this fragment

File: app/components/CardBet.tsx [hunk 5]
```diff
@@ -47,12 +52,24 @@ export const CardBet = ({ userBets, children }: CardBetProps) => {
...
+  const invalidMarketColleteralToRedeem =
+    parseFloat(formatEther(outcomeTokensTraded)) * 0.5;
+
+  const resultAmount =
+    !market.isClosed || isWinner
+      ? formatValueWithFixedDecimals(outcomeBalance, 2)
+      : market.isAnswerInvalid
+        ? formatValueWithFixedDecimals(invalidMarketColleteralToRedeem, 2)
+        : formatEtherWithFixedDecimals(collateralAmountSpent, 2);
+
```
[the remark this answers is not in the document]
Comment (Diogomartf) [reply app/components/CardBet.tsx:72]: Good suggestion, I think something like this is a step in the right direction.

**Comment by greenlucid** [comment]
Wording on the bottom is wrong, isn't it? If spent on bet was 30 WXDAI, but they receive 16.295 WXDAI from redeeming, they they got 32.59 shares, not 30 shares.
Also an user might have both YES and NO shares in the same market, no? Does that affect the rendering if that was the case?

**Comment by berteotti** [comment]
> Wording on the bottom is wrong, isn't it? If spent on bet was 30 WXDAI, but they receive 16.295 WXDAI from redeeming, they they got 32.59 shares, not 30 shares. 

You are right, user has 32.59 shares from 30 WXDAI he spent. Thanks for noticing.

> Also an user might have both YES and NO shares in the same market, no? Does that affect the rendering if that was the case?

We show each bet outcome independently.  So if you bet on YES and NO, we show 2 separate bets, showing share amount for each outcome.

File: app/components/CardBet.tsx [hunk 6]
```diff
@@ -14,13 +14,15 @@ import {
...
   tradesOutcomeBalance,
+  outcomeTokensTradedTotal,
   UserBets,
 } from '@/entities';
 import { redeemPositions } from '@/hooks/contracts';
 import { MarketThumbnail } from './MarketThumbnail';
 import { Skeleton } from './Skeleton';
 import { formatEtherWithFixedDecimals, formatValueWithFixedDecimals } from '@/utils';
 import { useTx } from '@/context';
+import { formatEther } from 'viem';
```
Comment (ElRodrigote) [inline app/components/CardBet.tsx:25]: [MINOR][NON-BLOCKING]
We can safely remove this import.

```suggestion
```

**Review by ElRodrigote — APPROVED** [review]

**Review by Diogomartf — APPROVED** [review]

**Pull request merged**